File: R/R/preprocess.R
#' Preprocess data by leaving it exactly the way it is.
#' 
#' This function is the same as the identity function. Anything passed
#' to it is returned untouched.
#'
#' @param .data A data.frame; the data to be analyzed.
#' @param categorical_variables A logical vector; each value TRUE indicates that column in the data.frame is a categorical variable. Defaults to NULL.
#' @return A list, containing one or two data.frames.
#' @family preprocess
#' @export
preprocess_identity <- function(.data, categorical_variables = NULL) {
  .data
}

#' Preprocess data by scaling it.
#' 
#' This function takes either a data.frame or a list of data.frames. In 
#' the event of the first, this function takes the dataset and will
#' scale each column that is not categorical such that that column
#' has zero mean and unit variance. In the event of the second, this 
#' function takes the training dataset and will identify the parameters 
#' needed to scale the training dataset such that each column that 
#' is not categorical has zero mean and unit variance, and then will 
#' apply those parameters to each column in the testing dataset 
#' that is not categorical.
#'
#' @param .data A data.frame; the data to be analyzed.
#' @param categorical_variables A logical vector; each value TRUE indicates that column in the data.frame is a categorical variable. Defaults to NULL.
#' @return A list, containing one or two data.frames.
#' @family preprocess
#' @export
preprocess_scale <- function(.data, categorical_variables = NULL) {
  mask <- categorical_variables
  if (length(.data) == 1) {
    # Handle case of list(X)
    X <- .data[["X"]]
    if (is.null(mask)) {
      # No categorical variables
      X_output <- data.frame(scale(X))
      output <- list(X = X_output)
    } else {
      # Categorical variables
      X_categorical <- X[, mask, drop = FALSE]
      X_numerical <- X[, !mask, drop = FALSE]
      X_standardized <- data.frame(scale(X_numerical))
      X_output <- cbind(X_categorical, X_standardized)
      output <- list(X = X_output)
    }
  } else if (length(.data) == 2) {
    # Handle case of list(X_train, X_test)
    X_train <- .data[["X_train"]]
    X_test <- .data[["X_test"]]
    if (is.null(mask)) {
      # No categorical variables
      # scale train
      X_train_scaled <- scale(X_train)
      scaled_center <- attr(X_train_scaled, "scaled:center")
      scaled_scale <- attr(X_train_scaled, "scaled:scale")
      
      # scale test
      X_test_scaled <- scale(X_test, scaled_center, scaled_scale)
      
      # output data
      X_train_output <- data.frame(X_train_scaled)
      X_test_output <- data.frame(X_test_scaled)
      output <- list(X_train = X_train_output, X_test = X_test_output)
    } else {
      # Categorical variables
      X_train_categorical <- X_train[, mask, drop = FALSE]
      X_train_numerical <- X_train[, !mask, drop = FALSE]
      X_test_categorical <- X_test[, mask, drop = FALSE]
      X_test_numerical <- X_test[, !mask, drop = FALSE]
      
      # scale train
      X_train_scaled <- scale(X_train_numerical)
      scaled_center <- attr(X_train_scaled, "scaled:center")
      scaled_scale <- attr(X_train_scaled, "scaled:scale")
      
      # scale test
      X_test_scaled <- scale(X_test_numerical, scaled_center, scaled_scale)
      
      # output data
      X_train_output <- cbind(X_train_categorical, data.frame(X_train_scaled))
      X_test_output <- cbind(X_test_categorical, data.frame(X_test_scaled))
      output <- list(X_train = X_train_output, X_test = X_test_output)
    }
  } else {
    stop("Value error: length(.data) is not 1 or 2.")
  }
  
  output
}

File: Python/easymlpy/preprocess.py
"""
Functions for preprocessing.
"""
import numpy as np
from sklearn.preprocessing import StandardScaler


__all__ = ['preprocess_identity', 'preprocess_scale']


def preprocess_identity(*data, categorical_variables=None):
    """    
    An identify function for preprocessing.

    Returns inputs without modifying them.

    :param data: array(s).
    :param categorical_variables: A list of strings representing the variables that are categorical.
    :return: array(s).
    """
    if len(data) == 1:
        output = data[0]
    elif len(data) == 2:
        output = data[0], data[1]
    else:
        raise ValueError
    return output


def preprocess_scale(*data, categorical_variables=None):
    """    
    A function for scaling data.

    Takes one or two arrays and scales them using a standard scaler.

    :param data: array(s).
    :param categorical_variables: A list of strings representing the variables that are categorical.
    :return: array(s).
    """
    sclr = StandardScaler()
    if len(data) == 1:
        X = data[0]
        if categorical_variables is not None:
            # import pdb; pdb.set_trace()
            X_categorical = X[:, categorical_variables]
            X_numerical = X[:, np.logical_not(categorical_variables)]
            X_standardized = sclr.fit_transform(X_numerical)
            output = np.concatenate([X_categorical, X_standardized], axis=1)
        else:
            output = sclr.fit_transform(X)
    elif len(data) == 2:
        X_train, X_test = data[0], data[1]
        if categorical_variables is not None:
            X_train_categorical = X_train[:, categorical_variables]
            X_train_numerical = X_train[:, np.logical_not(categorical_variables)]
            X_test_categorical = X_test[:, categorical_variables]
            X_test_numerical = X_test[:, np.logical_not(categorical_variables)]
            X_train_standardized = sclr.fit_transform(X_train_numerical)
            X_test_standardized = sclr.transform(X_test_numerical)
            X_train_output = np.concatenate([X_train_categorical, X_train_standardized], axis=1)
            X_test_output = np.concatenate([X_test_categorical, X_test_standardized], axis=1)
            output = X_train_output, X_test_output
        else:
            X_train_standardized = sclr.fit_transform(X_train)
            X_test_standardized = sclr.transform(X_test)
            output = X_train_standardized, X_test_standardized
    else:
        raise ValueError
    return output
